### ISSUE TITLE
lint suggestion: `suspicious_mixed_type_operation`

### ISSUE THREAD

**Issue by alv-around** [rust-lang/rust-clippy#17290]
### What it does

Bit-related functions return `u32` , this allows to compile some "suspicious" operations from mismatching calling types. which can leads to overflows like: 

```rust
let x = u8::BITS - 5u64.leading_zeros();

let x: u16 = 0b1111111111111111;
let _ = u8::BITS - x.trailing_ones();

let mask = 1u32 << 5u64.leading_zeros();
```

### Advantage

- Avoids overflows
- In the cases where it does not overflow, Improves readability 

### Drawbacks

The lint should be black list specific cases, which are clearly misleading. Otherwise, if the cases are kept to general there is the risk of triggering for false positives / true negatives.

### Example

```rust
let x = u8::BITS - 5u64.leading_zeros();

let x: u16 = 0b1111111111111111;
let _ = u8::BITS - x.trailing_ones();

let mask = 1u32 << 5u64.leading_zeros();
```

Could be written as:

```rust
let x = u8::BITS - 5u8.leading_zeros();

let x: u16 = 0b1111111111111111;
let _ = u16::BITS - x.trailing_ones();

let mask = 1u32 << 5u32.leading_zeros();
```

### Additional Context

This lint was inspired with the conversation on:  https://github.com/rust-lang/rust-clippy/pull/16902#discussion_r3412438731
